### ISSUE TITLE
How to change System.Private.CoreLib.dll find path?

### ISSUE THREAD

**Issue by CeSun** [dotnet/runtime#109633]
I am embedding coreclr using a custom host on a linux based mobile device.

When I initialized coreclr, coreclr prompted me that it could not find the path to System.Private.CoreLib.dll, and then it told me that the path to System.Private.CoreLib.dll was:
Path: /data/storage/el1/bundle/libs/arm64/System.Private.CoreLib.dll

The system I am using now has a sandbox mechanism. The system's software installer will only automatically copy native shared libraries to the "/data/storage/el1/bundle/libs/arm64/" directory. I do not have permission to automatically copy System.Private.CoreLib.dll there. Is there a way to modify the search directory for System.Private.CoreLib.dll?

**Comment by CeSun** [comment]
![Image](https://github.com/user-attachments/assets/0a38600d-4cbd-4451-bdf0-b3cf4e728d98)
![Image](https://github.com/user-attachments/assets/fde6cf89-e6ec-4614-9572-f68ae282d7f4)
![Image](https://github.com/user-attachments/assets/af5688ea-eef6-4bb9-930e-f9d9795048af)

**Comment by CeSun** [comment]
When I enter the system through the shell and manually copy System.Private.CoreLib.dll to the /data/storage/el1/bundle/libs/arm64/ directory, coreclr can be initialized successfully

**Comment by huoyaoyuan** [comment]
CoreLib is hardcoded to be next to coreclr because its tight relationship:

https://github.com/dotnet/runtime/blob/b567474c22d0766cf3a416592c39e4b633d015d9/src/coreclr/binder/assemblybindercommon.cpp#L253-L269

I'm afraid there's no viable workaround currently. Since you are writing custom host, single file bundling will not be an option.

**Comment by CeSun** [comment]
> CoreLib is hardcoded to be next to coreclr because its tight relationship:
> 
> [runtime/src/coreclr/binder/assemblybindercommon.cpp](https://github.com/dotnet/runtime/blob/b567474c22d0766cf3a416592c39e4b633d015d9/src/coreclr/binder/assemblybindercommon.cpp#L253-L269)
> 
> Lines 253 to 269 in [b567474](/dotnet/runtime/commit/b567474c22d0766cf3a416592c39e4b633d015d9)
> 
>  HRESULT AssemblyBinderCommon::BindToSystem(SString   &systemDirectory, 
>                                             Assembly **ppSystemAssembly) 
>  { 
>      HRESULT hr = S_OK; 
>   
>      _ASSERTE(ppSystemAssembly != NULL); 
>   
>      ReleaseHolder<Assembly> pSystemAssembly; 
>   
>      // System.Private.CoreLib.dll is expected to be found at one of the following locations: 
>      //   * Non-single-file app: In systemDirectory, beside coreclr.dll 
>      //   * Framework-dependent single-file app: In systemDirectory, beside coreclr.dll 
>      //   * Self-contained single-file app: Within the single-file bundle. 
>      // 
>      //   CoreLib path (sCoreLib): 
>      //   * Absolute path when looking for a file on disk 
>      //   * Bundle-relative path when looking within the single-file bundle. 
> I'm afraid there's no viable workaround currently. Since you are writing custom host, single file bundling will not be an option.

Bad news, I want to complain: not as good as mono.

**Comment by huoyaoyuan** [comment]
For the we-know-which platform that applies special processing for native executables/libraries, nativeaot would fit better since the compiled file is just a native object.

I'm also curious about how it handles multiple CLR copies from different applications. The versions of libcoreclr.so and System.Private.CoreLib must match exactly.

**Comment by driver1998** [comment]
I guess you can just add System.Private.CoreLib.dll inside the hap app package as well and let the system package manager handle it?

**Comment by huoyaoyuan** [comment]
Yeah, it's better to submit a request to change how package manager handles it.

**Comment by CeSun** [comment]
> Yeah, it's better to submit a request to change how package manager handles it.

It seems that it would be better to let coreclr adjust the mechanism of loading il to make it more suitable for systems such as android/harmonyos

**Comment by huoyaoyuan** [comment]
It's indeed hard. CoreLib is the specially handled instead of other managed assemblies. It's expected to be bundled closely together with coreclr.

**Comment by CeSun** [comment]
I personally think this is a design problem, mono can do it

**Comment by CeSun** [comment]
It looks like this environment variable can specify the directory: TRUSTED_PLATFORM_ASSEMBLIES
https://learn.microsoft.com/zh-cn/dotnet/core/dependency-loading/default-probing

**Comment by huoyaoyuan** [comment]
> It looks like this environment variable can specify the directory: TRUSTED_PLATFORM_ASSEMBLIES

No. This only affects the other `System.*` assemblies. CoreLib is very special and has a unique code path for loading.

**Comment by agocke** [comment]
So it sounds like the opposite approach could also work: we could make S.P.C a native assembly? Meaning the container file format would be the platform native executable format.

**Comment by CeSun** [comment]
> we could make S.P.C a native assembly?
I think this can be done.

**Comment by CeSun** [comment]
Any progress or plans? about "make S.P.C a native assembly"

**Comment by agocke** [comment]
No, this in the “future” milestone so it has no expected completion date.

**Comment by CeSun** [comment]
If this change is not a big deal, I hope to include it in the development plan as soon as possible. It affects whether coreclr can be embed in the mobile operating system.

**Comment by driver1998** [comment]
How does the package manager decide whether a file is "native shared libraries"? Just by filename or the file must be valid elf files? I wonder if simply renaming it `System.Private.CoreLib.so` will suffice.

**Comment by huoyaoyuan** [comment]
Package manager doesn't handle native library. CoreLib is not a native library. `.so` is only handled by the OS loader.
Loading a managed assembly and native library are totally unrelated.

**Comment by driver1998** [comment]
The package manager separates native libraries and other data content into different locations. If we can somehow trick the package manager into thinking `System.Private.CoreLib` is a native library, it'll place it next to `libcoreclr.so`.

**Comment by huoyaoyuan** [comment]
OK, you are referring to the package manager of the OS, not .NET.

Currently coreclr also hardcodes the file name of CoreLib. Changing the hardcoded file name may work.

Alternatively, if the CoreLib can be made in to R2R image in native ELF format like #120065, it can also be solved. I believe it's the most universal option since we are also doing so for iOS.

**Comment by AustinWise** [comment]
I think #128734 does what you want. It should ship in .NET 11.

**Comment by elinor-fung** [comment]
Per https://github.com/dotnet/runtime/issues/109633#issuecomment-4616118243, we added a new host-runtime property `SYSTEM_CORELIB_DIRECTORY` that allows setting the directory in .NET 11.
